File: cpp/D6.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <regex>
#include <unordered_set>
#include <unordered_map>
#include "time.h"
#include "text.h"

using namespace std;

vector<vector<string>> read_input() {
    ifstream input("inputs\\D6.input");
    vector<vector<string>> blocks = split_blocks(&input);
    vector<vector<string>> groups;
    for (vector<string> block : blocks) {
        vector<string> group;
        for (string person : block) {
            group.push_back(person);
        }
        groups.push_back(group);
    }
    
    return groups;
}

int P1(vector<vector<string>> groups) {
    int count = 0;
    for (vector<string> group : groups) {
        unordered_set<char> questions;
        for (string person : group)
        for (char question : person)
            questions.insert(question);
        count += questions.size();
    }
    return count;
}

int P2(vector<vector<string>> groups) {
    int count = 0;
    for (vector<string> group : groups) {
        unordered_map<char,int> questions;
        for (string person : group)
        for (char question : person)
            questions[question]++;

        for (auto kvp : questions)
            if (kvp.second == group.size())
                count++;
    }
    return count;
}

int main() {
    run_with_time<int>([](void) { 
        return P1(read_input()); 
    });

    run_with_time<int>([](void) { 
        return P2(read_input()); 
    });
}

File: ts/D6.ts
import { readFileSync } from 'fs';
import { timeAndPrint } from './time-and-print';

function parseInput(): string[][] {
    const groups = [];
    const content = readFileSync('../inputs/D6.input','utf8');
    for (let section of content.split("\r\n\r\n")) {
        const group = [];
        for (let line of section.split("\r\n")) {
            group.push(line);
        }
        groups.push(group);
    }
    return groups;
}

function countAnsweredQuestions(group: string[]): number {
    const set = new Set<string>();
    for (let person of group)
        for (let question of person)
            set.add(question);
    return set.size;
}

function p1() {
    return parseInput()
        .map(countAnsweredQuestions)
        .reduce((a, b) => a + b);
}

function countUnamiousQuestions(group: string[]): number {
    const hash: { [key: string]: number } = {};
    for (let person of group)
        for (let question of person)
            hash[question] ? hash[question]++ : hash[question] = 1;

    return Object.keys(hash)
        .filter(k => hash[k] == group.length)
        .length;
}

function p2() {
    return parseInput()
        .map(countUnamiousQuestions)
        .reduce((a,b) => a + b);
}

timeAndPrint(p1);
timeAndPrint(p2);